### PR TITLE
[IMP] stock: safe reservation in high concurrency

### PR DESCRIPTION
**Pull request by nseinlet**
Idea is to use same mechanism as in _update_available_quantity, update 
if possible, or create a new one and have _merge_quants to solve the 
duplication.
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by nseinlet** [comment]
@stenyx-de

File: addons/stock/models/stock_quant.py [hunk 1]
```diff
@@ -292,6 +292,29 @@ def _update_reserved_quantity(self, product_id, location_id, quantity, lot_id=No
...
         return reserved_quants
 
+    @api.multi
+    def _safe_reserved_qty_upd(self, qty):
+        for quant in self:
+            try:
+                with self._cr.savepoint():
+                    self._cr.execute("SELECT 1 FROM stock_quant WHERE id = %s FOR UPDATE NOWAIT", [quant.id], log_exceptions=False)
+                    quant.write({'reserved_quantity': quant.reserved_quantity + qty})
+                    break
```
Comment (fmdl) [inline addons/stock/models/stock_quant.py]: why `break` ?

**Comment by nseinlet** [comment]
Doesn't work, as reservation always succeed, even when it should fail. 20 concurrent process can reserve sucessfully, even if you only have 10 in stock.

**Comment by stenyx-de** [comment]
Is there any other way to maybe "avoid" serialization failures in the quants?
We are still very much confronted with this issue.

Output of our database log, to show how often it occurs.
(Of course, it occurs a few times more and a few times less often because it is a problem with parallelism.)
```
2019-04-30 06:44:05.094 UTC [75786] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:05.094 UTC [75786] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28428.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
2019-04-30 06:44:05.094 UTC [37500] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:05.094 UTC [37500] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28428.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
2019-04-30 06:44:06.699 UTC [37500] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:06.699 UTC [37500] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28430.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
2019-04-30 06:44:06.699 UTC [75786] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:06.699 UTC [75786] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28430.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
2019-04-30 06:44:08.417 UTC [75786] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:08.417 UTC [75786] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28431.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
2019-04-30 06:44:08.417 UTC [37500] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:08.417 UTC [37500] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28431.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
2019-04-30 06:44:09.972 UTC [37500] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:09.972 UTC [37500] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28432.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
2019-04-30 06:44:11.758 UTC [37500] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:11.758 UTC [37500] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28434.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
2019-04-30 06:44:16.577 UTC [38854] odoo@odoo ERROR:  could not serialize access due to concurrent update
2019-04-30 06:44:16.577 UTC [38854] odoo@odoo STATEMENT:  UPDATE "stock_quant" SET "reserved_quantity"=28435.0,"write_uid"=1,"write_date"=(now() at time zone 'UTC') WHERE id IN (3500806)
```

**Comment by stenyx-de** [comment]
@nseinlet is there anything new on this?

**Comment by nseinlet** [comment]
We cannot make the reservation always succeed. In your case, you've enough stock, but let's say you only have qty=1, and 2 reservations, one should succeed, and the other should fail.

**Comment by stenyx-de** [comment]
Okay, but it is a serious problem our users face when they are working and there is some rush at this day.
We get compaints from our users that the issue is occurring and the only thing we can currently do is to tell them, that it is a problem due to system load; but this isnt a gratifying answer.

**Pull request closed without merging**